pkg/antctl/raw/supportbundle/command.go

```diff
@@ -692,9 +702,164 @@ func processResults(resultMap map[string]error, dir string) error {
+
+	if err := func() error {
+		controllerInfo, err := antreaClientset.CrdV1beta1().AntreaControllerInfos().Get(ctx, "antrea-controller", metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		data, err := yaml.Marshal(controllerInfo)
+		if err != nil {
+			return err
+		}
+		if err := afero.WriteFile(defaultFS, filepath.Join(dir, "controllerinfo"), data, 0644); err != nil {
+			return err
+		}
+		return nil
+	}(); err != nil {
+		errors = append(errors, err)
+	}
+
+	pods, err := k8sClient.CoreV1().Pods("kube-system").List(ctx, metav1.ListOptions{
+		ResourceVersion: "0",
+		LabelSelector:   "app=antrea,component=antrea-controller",
+	})
+	errors = append(errors, err)
+	if err == nil {
+		for _, pod := range pods.Items {
+			err = downloadPodLogs(ctx, k8sClient, pod.Namespace, pod.Name, k8s.GetPodContainerNames(&pod), tmpDir)
+			errors = append(errors, err)
+			gzFileName := filepath.Join(dir, "controller_"+pod.Spec.NodeName+".tar.gz")
+			f, err := defaultFS.Create(gzFileName)
+			if err != nil {
+				errors = append(errors, err)
+			} else {
+				defer f.Close()
+				_, err := compress.PackDir(defaultFS, tmpDir, f)
+				errors = append(errors, err)
+			}
+		}
+	}
+	return utilerror.NewAggregate(errors)
+}
+
+func downloadFallbackAgentBundleFromKubernetes(ctx context.Context, antreaClientset antrea.Interface, k8sClient kubernetes.Interface, failedNodes []string, dir string) error {
+	agentInfoList, err := antreaClientset.CrdV1beta1().AntreaAgentInfos().List(ctx, metav1.ListOptions{ResourceVersion: "0"})
+	if err != nil {
+		return err
+	}
+
+	agentInfoMap := map[string]v1beta1.AntreaAgentInfo{}
+	for _, agentInfo := range agentInfoList.Items {
+		agentInfoMap[agentInfo.Name] = agentInfo
+	}
+	pods, err := k8sClient.CoreV1().Pods("kube-system").List(ctx, metav1.ListOptions{
+		ResourceVersion: "0",
+		LabelSelector:   "app=antrea,component=antrea-agent",
+	})
+	if err != nil {
+		return err
+	}
+	failedNodeSet := sets.NewString(failedNodes...)
+	var errors []error
+	for _, pod := range pods.Items {
+		if !failedNodeSet.Has(pod.Spec.NodeName) {
+			continue
+		}
+		tmpDir, err := afero.TempDir(defaultFS, "", "bundle_tmp_")
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		defer defaultFS.RemoveAll(tmpDir)
+
+		if agentInfo, ok := agentInfoMap[pod.Spec.NodeName]; ok {
+			data, err := yaml.Marshal(agentInfo)
+			errors = append(errors, err)
+			if err == nil {
+				err = afero.WriteFile(defaultFS, filepath.Join(tmpDir, "agentinfo"), data, 0644)
+				errors = append(errors, err)
+			}
+		}
+		err = downloadPodLogs(ctx, k8sClient, pod.Namespace, pod.Name, k8s.GetPodContainerNames(&pod), tmpDir)
+		errors = append(errors, err)
+		gzFileName := filepath.Join(dir, "agent_"+pod.Spec.NodeName+".tar.gz")
+		f, err := defaultFS.Create(gzFileName)
+		if err != nil {
+			errors = append(errors, err)
+		} else {
+			defer f.Close()
+			_, err := compress.PackDir(defaultFS, tmpDir, f)
+			errors = append(errors, err)
+		}
```

Comment:
maybe find a way to write a helper function that can be used by both `downloadFallbackControllerBundleFromKubernetes` and `downloadFallbackAgentBundleFromKubernetes`